core/drivers/zynqmp_pm.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021 Foundries.io Ltd
+ */
+
+#include <arm.h>
+#include <drivers/zynqmp_pm.h>
+#include <kernel/cache_helpers.h>
+#include <mm/core_memprot.h>
+#include <smccc.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+#include <utee_defines.h>
+
+#define EFUSE_ACCESS_SMC	0xC2000035
+#define EFUSE_NOT_ENABLED	29
+
+enum efuse_op {EFUSE_READ = 0, EFUSE_WRITE = 1};
+
+/*
+ * Stores all required details to
+ * read/write efuse memory.
+ * @src:        paddr_t of the buffer to store the data to be write/read
+ *              buffer should be cache aligned
+ * @size:       number of 32-bit words to be read/written
+ * @offset:     offset to be read/written
+ * @flag:       EFUSE_READ  - represents efuse read operation
+ *              EFUSE_WRITE - represents efuse write operation
+ * @pufuserfuse:0 - represents non-puf efuses, offset is used for read/write
+ *              1 - represents puf user fuse row number.
+ */
+struct xilinx_efuse {
+	uint64_t src;
+	uint32_t size;
+	uint32_t offset;
+	uint32_t flag;
+	uint32_t pufuserfuse;
+};
+
+static uint32_t zynqmp_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				uint32_t *payload)
+{
+	struct smccc_res res = { };
+
+	smccc_smc(pm_api_id,
+		  ((uint64_t)arg1 << 32) | arg0,
+		  ((uint64_t)arg3 << 32) | arg2,
+		  0, 0, 0, 0, 0, &res);
+
+	if (payload && !res.a0)
+		*payload = res.a0 >> 32;
+
+	return res.a0;
+}
+
+static void *alloc_aligned_zeroed(size_t sz)
+{
+	size_t alloc_size = 0;
+	size_t cacheline_size = 0;
+	void *ptr = NULL;
+
+	alloc_size = sz;
+	cacheline_size = dcache_get_line_size();
+
+	if (ADD_OVERFLOW(alloc_size, ROUNDUP(alloc_size, cacheline_size),
+			 &alloc_size))
+		return NULL;
+
+	ptr = memalign(cacheline_size, alloc_size);
+	if (!ptr)
+		return NULL;
+
+	memset(ptr, 0, alloc_size);
+
+	return ptr;
+}
+
+static TEE_Result efuse_op(enum efuse_op op, uint8_t *buf, size_t sz,
+			   uint32_t efuse_offset, bool puf)
+{
+	paddr_t addr = 0;
+	void *buf_aligned = NULL;
+	uint32_t res = 0;
+	TEE_Result ret = TEE_SUCCESS;
+	struct xilinx_efuse *efuse = NULL;
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	buf_aligned = alloc_aligned_zeroed(sz);
+	if (!buf_aligned)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	efuse = alloc_aligned_zeroed(sizeof(struct xilinx_efuse));
+	if (!efuse) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto alloc_error;
+	}
+
+	memcpy(buf_aligned, buf, sz);
+
+	efuse->src = (uint64_t)virt_to_phys(buf_aligned);
+	efuse->size = sz / sizeof(uint32_t);
+	efuse->offset = efuse_offset;
+	efuse->flag = op;
+	efuse->pufuserfuse = puf;
+
+	cache_operation(TEE_CACHEFLUSH, (void *)buf_aligned, sz);
```

Comment:
It seems a bit strange that this isn't handled by the firmware at EL3. If this SMC interface already is written in stone I guess that's no way around it, but otherwise I'd recommend fixing it.

By the way, it should be enough to clean the cache entries here.

Reply:
> By the way, it should be enough to clean the cache entries here.
Right, fixed, thanks!